# Trees for phylogenetic analyses

`chr_W_Host.SNP.DP3-AC2-MQ40.min4.phy.contree`: used for plotting collapsed matrilineal phylogeny, e.g. Fig 1A. 

`chr_MT_Host.SNP.DP3-AC2-MQ40.min4.phy.contree`: used for plotting collapsed matrilineal phylogeny. 

`chr_W.trees.ann`: SNAPP annotated trees of chrW. 

`chr_W_HostOG...`: These 4 annotated trees show the BEAST time dated phylogenies with difference sensitivities (e.g. constant or exponential). 

`chr_MT_All.SNP.DP3-AC2-MQ40.min4.phy.contree`: used for classifying all samples into haplogroups 

`chr_W_All.SNP.DP3-AC2-MQ40.min4.phy.contree`: same as above, except only higher quality W chromosome for checking. 